### PR TITLE
Add query profiler based on resolution tree

### PR DESCRIPTION
**Pull request by kasper-piskorski**
## What is the goal of this PR?
To add a better way of examining query execution with reasoning by providing a way of outputting the resolution tree in a graphical-friendly form. The tree contains both answer and processing time per node information.

## What are the changes implemented in this PR?
With logging set to `TRACE`, we now explicitly store the full ResolutionTree.
Once the query is finished executing, the tree is then output in Graph Description Laguage format (https://en.wikipedia.org/wiki/DOT_(graph_description_language) to the log directory.

### CODE REVIEW

File: graql/reasoner/tree/TreeWriter.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+public class TreeWriter {
```
Comment (flyingsilverfin) [inline graql/reasoner/tree/TreeWriter.java]: class comments on what this is andwhy

File: graql/reasoner/tree/ResolutionTree.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ *
+ */
+public class ResolutionTree {
```
Comment (flyingsilverfin) [inline graql/reasoner/tree/ResolutionTree.java]: comment on class

File: graql/reasoner/tree/ResolutionTree.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    public Set<Node> getNodes(){
+        return new HashSet<>(mapping.values());
+    }
+
+    public void clear(){
+        mapping.clear();
+    }
+
+    public void updateTree(ResolutionState state) {
```
Comment (flyingsilverfin) [inline graql/reasoner/tree/ResolutionTree.java]: Can we rename this to be  abit more descriptive? like `recordNewState` or `addResolutionState`?

File: graql/reasoner/ResolutionIterator.java [hunk 4]
```diff
@@ -115,19 +132,22 @@ public boolean hasNext() {
...
                 return hasNext();
             }
         }
 
         MultilevelSemanticCache queryCache = CacheCasting.queryCacheCast(this.queryCache);
-
         subGoals.forEach(queryCache::ackCompleteness);
         queryCache.propagateAnswers();
 
+        if (LOG.isTraceEnabled()) logTree.outputToFile(Paths.get("/tmp/query.profile"));
```
Comment (flyingsilverfin) [inline graql/reasoner/ResolutionIterator.java]: I wonder if we can find a better place to put this file, would break on windows atm right

Comment (kasper-piskorski) [reply graql/reasoner/ResolutionIterator.java]: ideally it should be in the log dir. Not sure how to access it.

Comment (flyingsilverfin) [reply graql/reasoner/ResolutionIterator.java]: ask @vmax

Comment (kasper-piskorski) [reply graql/reasoner/ResolutionIterator.java]: changed

**Pull request merged**